sdk/ml/azure-ai-ml/azure/ai/ml/entities/_inputs_outputs/external_data.py

```diff
@@ -0,0 +1,183 @@
+        Details will be provided in the error message.
+    """
+
+    _EMPTY = Parameter.empty
+
+    def __init__(
+        self,
+        *,
+        type: Optional[str] = ExternalDataType.DATABASE,  # pylint: disable=redefined-builtin
+        query: Optional[str] = None,
+        table_name: Optional[str] = None,
+        stored_procedure: Optional[str] = None,
+        stored_procedure_params: Optional[List[dict]] = None,
+        connection: Optional[str] = None,
+    ):
+        # As an annotation, it is not allowed to initialize the name.
+        # The name will be updated by the annotated variable name.
+        self.type = type
+        self.name = None
+        self.connection = self._to_data_binding(connection)
+        self.query = self._to_data_binding(query)
+        self.table_name = self._to_data_binding(table_name)
+        self.stored_procedure = stored_procedure
+        self.stored_procedure_params = stored_procedure_params
+
+    def _to_data_binding(self, value):  # pylint: disable=no-self-use
+        if "_data_binding" in type(value).__dict__:
+            return value._data_binding()  # pylint: disable=protected-access
+        return value
+
+    def _to_dict(self, remove_name=True):
+        """Convert the Source object to a dict."""
+        keys = ["name", "type", "query", "stored_procedure", "stored_procedure_params", "connection", "table_name"]
+        if remove_name:
+            keys.remove("name")
+        result = {key: getattr(self, key) for key in keys}
+        return _remove_empty_values(result)
+
+    def _to_rest_object(self) -> Dict:
+        # this is for component rest object when using Source as component inputs, as for job input usage,
+        # rest object is generated by extracting Source's properties, see details in to_rest_dataset_literal_inputs()
+        result = self._to_dict()
+        return result
+
+    def _update_name(self, name):
+        self.name = name
+
+    @classmethod
+    def _from_rest_object(cls, obj: Dict) -> "Database":
+        return Database(**obj)
+
+    @property
+    def stored_procedure_params(self) -> Optional[StoredProcedureParameter]:
+        """Compute Resource configuration for the job."""
+
+        return self._stored_procedure_params
+
+    @stored_procedure_params.setter
+    def stored_procedure_params(self, value: Union[Dict[str, str], StoredProcedureParameter, None]):
+        if value is None:
+            self._stored_procedure_params = value
+        else:
+            if not isinstance(value, list):
+                value = [value]
+            for index, item in enumerate(value):
+                if isinstance(item, dict):
+                    value[index] = StoredProcedureParameter(**item)
+            self._stored_procedure_params = value
+
+
+class FileSystem(DictMixin, RestTranslatableMixin):  # pylint: disable=too-many-instance-attributes
+    """Define a file system class of a DataTransfer Component or Job.
```

Comment:
might improve the docstring, e.g. add some examples.

Reply:
added examples